autode/wrappers/G09.py

```diff
@@ -36,180 +43,261 @@ def modify_keywords_for_point_charges(keywords):
+def rerun_angle_failure(calc):
+    """
+    Gaussian will sometimes encounter a 180 degree angle and crash. This
+    function performs a few geometry optimisation cycles in cartesian
+    coordinates then switches back to internals
 
-        for keyword in keywords:
+    Arguments:
+        calc (autode.calculation.Calculation):
 
-            if 'opt' in keyword.lower():
-                calc.opt = True
+    Returns:
+        (autode.calculation.Calculation):
+    """
+    cart_calc = deepcopy(calc)
 
-                if calc.n_atoms == 1:
-                    logger.warning('Cannot do an optimisation for a single atom')
-                    keywords.remove(keyword)
+    # Iterate through a copied set of keywords
+    for keyword in copy(cart_calc.input.keywords):
+        if keyword.lower().startswith('geom'):
+            cart_calc.input.keywords.remove(keyword)
 
-        if calc.point_charges:
-            modify_keywords_for_point_charges(keywords)
+        elif keyword.lower().startswith('opt'):
+            options = []
+            if '=(' in keyword:
+                # get the individual options
+                options = [option.lower().strip()
+                           for option in keyword[5:-1].split(',')]
 
-        # By default perform all optimisations without symmetry
-        if calc.opt and not any(k.lower() == 'nosymm' for k in keywords):
-            keywords.append('NoSymm')
+                for option in options:
+                    if (option.startswith('maxcycles')
+                            or option.startswith('maxstep')):
+                        options.remove(option)
 
-        with open(calc.input_filename, 'w') as inp_file:
-            print(f'%mem={calc.max_core_mb}MB', file=inp_file)
+            elif '=' in keyword:
+                options = [keyword[4:]]
+            options += ['maxcycles=3', 'maxstep=1', 'cartesian']
+
+            new_keyword = f'Opt=({", ".join(options)})'
+            cart_calc.input.keywords.remove(keyword)
+            cart_calc.input.keywords.append(new_keyword)
+
+    # Generate the new calculation and run
+    cart_calc.name += '_cartesian'
+    cart_calc.molecule.atoms = calc.get_final_atoms()
+    cart_calc.molecule.constraints = Constraints(distance=None, cartesian=None)
+    cart_calc.input.added_internals = None
+    cart_calc.output = CalculationOutput()
+    cart_calc.run()
+
+    if not cart_calc.terminated_normally():
+        logger.warning('Cartesian calculation did not converge')
+        return None
+
+    logger.info('Returning to internal coordinates')
+
+    # Reset the required parameters for the new calculation
+    fixed_calc = deepcopy(calc)
+    fixed_calc.name += '_internal'
+    fixed_calc.molecule.atoms = cart_calc.get_final_atoms()
+    fixed_calc.output = CalculationOutput()
+    fixed_calc.run()
+
+    return fixed_calc
+
+
+class G09(ElectronicStructureMethod):
+
+    def generate_input(self, calc, molecule):
```

Comment:
it seems like we're halfway to having a "generate_input()" in base instead of in each wrapper

Reply:
but not quite..